source/s3_buffer_pool.c

```diff
@@ -101,151 +142,178 @@ void aws_s3_buffer_pool_destroy(struct aws_s3_buffer_pool *buffer_pool) {
     aws_mem_release(base, buffer_pool);
 }
 
-void aws_s3_buffer_pool_trim(struct aws_s3_buffer_pool *buffer_pool) {
-    size_t swap_count = 0;
-
+void s_buffer_pool_trim_synced(struct aws_s3_buffer_pool *buffer_pool) {
     for (size_t i = 0; i < aws_array_list_length(&buffer_pool->blocks); ++i) {
         struct s3_buffer_pool_block *block;
         aws_array_list_get_at_ptr(&buffer_pool->blocks, (void **)&block, i);
 
         if (block->alloc_count == 0) {
-            size_t swap_loc = aws_array_list_length(&buffer_pool->blocks) - 1 - swap_count;
-            if (swap_loc == i) {
-                break;
-            } else {
-                aws_array_list_swap(&buffer_pool->blocks, i, swap_loc);
-            }
+            aws_mem_release(buffer_pool->base_allocator, block->block_ptr);
+            aws_array_list_erase(&buffer_pool->blocks, i);
```

Comment:
don't make any changes unless we actually find that this is a bottleneck some day... just sharing a thought:
if this were a linked-list, erase would be O(1) instead of O(N)
BUT then we'd need to allocate each block struct, instead of letting it live by-value in an array-list 🤷‍♀️

Reply:
Agree. i was somewhat concerned about O(N) and went with a swapping approach to reduce complexity of overall trim. That was probably premature optimization and from some initial testing it does not seem to affect perf too much.